atc/worker/container.go

```diff
@@ -138,3 +146,110 @@ func (container *gardenWorkerContainer) initializeVolumes(
 
 	return nil
 }
+
+// TODO (runtime/#4910): this needs to be modified to not be resource specific
+// 		the stdout of the run() is expected to be of json format
+//      this will break if used with task_step as it does not
+//		print out json
+func (container *gardenWorkerContainer) RunScript(
+	ctx context.Context,
+	path string,
+	args []string,
+	input []byte,
+	output interface{},
+	logDest io.Writer,
+	recoverable bool,
+) error {
+	if recoverable {
+		result, _ := container.Properties()
+		code := result[runtime.ResourceResultPropertyName]
+		if code != "" {
+			return json.Unmarshal([]byte(code), &output)
+		}
+	}
+
+	stdout := new(bytes.Buffer)
+	stderr := new(bytes.Buffer)
+
+	processIO := garden.ProcessIO{
+		Stdin:  bytes.NewBuffer(input),
+		Stdout: stdout,
+	}
+
+	if logDest != nil {
+		processIO.Stderr = logDest
+	} else {
+		processIO.Stderr = stderr
+	}
+
+	var process garden.Process
+	var err error
+	if recoverable {
+		process, err = container.Attach(ctx, runtime.ResourceProcessID, processIO)
+		if err != nil {
+			process, err = container.Run(
+				ctx,
+				garden.ProcessSpec{
+					ID:   runtime.ResourceProcessID,
+					Path: path,
+					Args: args,
+				}, processIO)
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		process, err = container.Run(ctx, garden.ProcessSpec{
+			Path: path,
+			Args: args,
+		}, processIO)
+		if err != nil {
+			return err
+		}
+	}
+
+	processExited := make(chan struct{})
+
+	var processStatus int
+	var processErr error
+
+	go func() {
+		processStatus, processErr = process.Wait()
+		close(processExited)
+	}()
+
+	select {
+	case <-processExited:
+		if processErr != nil {
+			return processErr
+		}
+
+		if processStatus != 0 {
+			return runtime.ErrResourceScriptFailed{
+				Path:       path,
+				Args:       args,
+				ExitStatus: processStatus,
+
+				Stderr: stderr.String(),
+			}
+		}
+
+		if recoverable {
+			err := container.SetProperty(runtime.ResourceResultPropertyName, stdout.String())
+			if err != nil {
+				return err
+			}
+		}
+
+		err := json.Unmarshal(stdout.Bytes(), output)
+		if err != nil {
+			return fmt.Errorf("%s\n\nwhen parsing resource response:\n\n%s", err, stdout.String())
+		}
+		return err
+
+	case <-ctx.Done():
+		container.Stop(false)
```

Comment:
from the Garden interface, `container.Stop(kill)` returns an error. Are we ignoring it on purpose? (I guess so).

If so, I wonder if we should signal that, e.g.:

```
_ = container.Stop(false)
```

thanks!